### PR TITLE
 [perceval] Add poetry support

### PR DESCRIPTION
**Pull request by vchrombie**
This commit adds the files required to add poetry support to perceval.

This update is required to start using the Bitergia/release-tools for automating the GrimoireLab toolset releases.

### CODE REVIEW

**Comment by vchrombie** [comment]
I took reference from [grimoirelab-toolkit](https://github.com/chaoss/grimoirelab-toolkit/) and [release-tools](https://github.com/Bitergia/release-tools) while preparing this PR.

I followed the [instructions](https://github.com/chaoss/grimoirelab-toolkit/pull/27#issuecomment-668488818) mentioned by @mafesan and it worked perfectly fine for me. 

But, I faced one error while executing the perceval script
```
(venv) $ perceval git https://github.com/chaoss/grimoirelab-perceval
Traceback (most recent call last):
  File "/home/p0tt3r/dev/dump/grimoirelab-perceval/venv/bin/perceval", line 2, in <module>
    from bin.perceval import main
ModuleNotFoundError: No module named 'bin.perceval'
```
To resolve this, I renamed the perceval script file in the bin folder from `bin/perceval` to `bin/perceval.py` and executed `poetry install`. Later the command worked perfectly fine.
```
(venv) $ perceval git https://github.com/chaoss/grimoirelab-perceval
[2020-10-14 14:10:27,152] - Sir Perceval is on his quest.
[2020-10-14 14:10:27,155] - Fetching commits: 'https://github.com/chaoss/grimoirelab-perceval' git repository from 1970-01-01 00:00:00+00:00 to 2100-01-01 00:00:00+00:00; all branches
...
```

I think the files must be python (end in .py but not executable files) in order to add them as poetry scripts, not sure tho. I need to check if we can overcome this issue in anyway.
```
[tool.poetry.scripts]
perceval = 'bin.perceval:main'
```
@sduenas, do you have any comments?

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+[tool.poetry.dependencies]
+python = "^3.6"
+python-dateutil = ">=2.6.0"
+requests = ">=2.7.0"
+beautifulsoup4 = ">=4.3.2"
+feedparser = ">=5.1.3"
+dulwich = ">=0.20.0"
+urllib3 = ">=1.22"
+grimoirelab-toolkit = {git = "https://github.com/chaoss/grimoirelab-toolkit.git"}
```
Comment (sduenas) [inline pyproject.toml:44]: This should point to the last version of `grimoirelab-toolkit` instead of to the repository.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+]
+
+packages = [
+    { include = "perceval" },
+    { include = "tests", format = "sdist" },
+]
+
+classifiers = [
+   "Development Status :: 4 - Beta",
+   "Intended Audience :: Developers",
```
Comment (sduenas) [inline pyproject.toml:27]: We can also add `Intended Audience :: Science/Research`.

**Review by sduenas — CHANGES_REQUESTED** [review]
Please check my comments.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+   "Topic :: Software Development",
+   "License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)",
+   "Programming Language :: Python :: 3"
+]
+
+[tool.poetry.urls]
+"Bug Tracker" = "https://github.com/chaoss/grimoirelab-perceval/issues"
+
+[tool.poetry.scripts]
+perceval = 'bin.perceval:main'
```
Comment (vchrombie) [inline pyproject.toml:38]: To support this feature, we need to change the perceval script from `bin/perceval` to `bin/perceval.py`. If we don't need this change, we can remove these two lines.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+[tool.poetry.dependencies]
+python = "^3.6"
```
Comment (vchrombie) [inline pyproject.toml:41]: Do you want me to mention the various versions here as we did in ELK [pyproject.toml#L41](https://github.com/chaoss/grimoirelab-elk/blob/master/pyproject.toml#L41)?
I feel this is enough, but what are your thoughts on this?

**Review by vchrombie — COMMENTED** [review]
Hi @sduenas 
I have updated the PR, you can have a look again.

Also, I had a few questions. I have addressed them in the inline comments.

File: pyproject.toml (hunk 1 above)
Comment (vchrombie) [reply pyproject.toml]: How does it look now?

File: pyproject.toml (hunk 3 above)
Comment (sduenas) [reply pyproject.toml:38]: I'd like to keep perceval script as it is. No `.py` extensions.

File: pyproject.toml (hunk 4 above)
Comment (sduenas) [reply pyproject.toml:41]: I think that's enough. It will work for `3.x`.

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+[tool.poetry]
+name = "perceval"
+version = "0.17.1"
+description = "Fetch data from software repositories"
```
Comment (sduenas) [inline pyproject.toml:4]: I'd like to change this to `Send Sir Perceval on a quest to fetch and gather data from software repositories.`.

**Review by sduenas — COMMENTED** [review]
Please check my comments.

File: pyproject.toml (hunk 3 above)
Comment (vchrombie) [reply pyproject.toml:38]: Okay, I'm removing the `[tool.poetry.scripts]` as it doesn't support in this case. :+1:

**Comment by vchrombie** [comment]
Hi @sduenas, I updated the PR as per your comments. 
Let me know if we need to add anything more.

Do you think we should add any documentation?
I understand that we need poetry support only for release-tools but I was thinking if there is anything to be added in the README.md file or anywhere.
Thanks.

**Comment by sduenas** [comment]
> Hi @sduenas, I updated the PR as per your comments.
> Let me know if we need to add anything more.
> 
> Do you think we should add any documentation?
> I understand that we need poetry support only for release-tools but I was thinking if there is anything to be added in the README.md file or anywhere.
> Thanks.

I think that will be the next step when we're ready to use the `release-tools` with `perceval`. First, we need to have the other perceval repositories working with it and then we can talk about the rules for contributing. What do you think?

**Comment by vchrombie** [comment]
> I think that will be the next step when we're ready to use the `release-tools` with `perceval`. First, we need to have the other perceval repositories working with it and then we can talk about the rules for contributing. What do you think?

Perfect! Now that I have this PR ready, I'll open the same PRs in other perceval repositories too.

**Comment by vchrombie** [comment]
Hi @sduenas, the PRs are ready.

- [x] https://github.com/chaoss/grimoirelab-perceval-opnfv/pull/33
- [x] https://github.com/chaoss/grimoirelab-perceval-mozilla/pull/56
- [x] https://github.com/chaoss/grimoirelab-perceval-puppet/pull/28
- [x] https://github.com/Bitergia/grimoirelab-perceval-finos/pull/19
- [x] https://github.com/chaoss/grimoirelab-perceval-weblate/pull/4

These 6 PRs (5+current) are the ones for perceval.
Feel free to review when you have time.
Thanks.

**Pull request merged**